### PR TITLE
Install ansible-test

### PR DESCRIPTION
**Pull request by abadger**
Modify the install script to install ansible-test and its supporting
code.  Alternative to #60701 that doesn't change package_dir ansible for
fear that it might regress https://github.com/ansible/ansible/issues/10437
and uses a function to calculate the package_data

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME
setup.py

### CODE REVIEW

**Comment by abadger** [comment]
I think this now has all the requested features.  A few more things to do:

* One thing from the review of https://github.com/ansible/ansible/pull/60701 that still needs to be added: See if test/sanity/code-smell/package-data.py can be modified to check for files that aren't being installed.

* Manually check that all the files are being included in the sdist and that installation from the sdist is complete.

**Comment by abadger** [comment]
@sivel @nitzmahone I've added the package_data tests to a separate commit here (in case we want the other PR, the test changes can just be stolen from here).  Something hat occured to me as I was doing this, though.  Should we sdist first and then install from the sdist to check for missing files?  Various parts of the setup.py install step are simply installing all of the files in the tree that match certain patterns.  If the files aren't added to the sdist, then the install from the sdist won't contain the files even though an install from the git tree did.

**Comment by abadger** [comment]
I did a manual check of an install from sdist.  All files are present.  this seems ready for review.

File: MANIFEST.in [hunk 1]
```diff
@@ -4,19 +4,32 @@ include SYMLINK_CACHE.json
 include requirements.txt
 include shippable.yml
 include bin/ansible-test
+recursive-include docs *
+exclude docs/docsite/rst_warnings
+recursive-exclude docs/docsite/_build *
+recursive-exclude docs/docsite/extensions *.pyc *.pyo
```
Comment (mattclay) [inline MANIFEST.in]: This path doesn't exist. Did you mean this?

```suggestion
recursive-exclude docs/docsite/_extensions *.pyc *.pyo
```

File: MANIFEST.in [hunk 2]
```diff
@@ -4,19 +4,32 @@ include SYMLINK_CACHE.json
...
-recursive-include lib/ansible/module_utils/powershell *
-recursive-include lib/ansible/modules *
-recursive-include lib/ansible/galaxy/data *
-recursive-include docs *
+recursive-include lib/ansible/executor/powershell *.ps1
+recursive-include lib/ansible/module_utils/csharp *.cs
+recursive-include lib/ansible/module_utils/powershell *.psm1
+recursive-include lib/ansible/modules/windows *.ps1
+recursive-include lib/ansible/galaxy/data *.yml *.j2 README.md ansible.cfg inventory .git_keep
+recursive-include lib/ansible/config/ *.yml
```
Comment (mattclay) [inline MANIFEST.in]: This may just be cosmetic, but why does this directory have a trailing `/` when none of the others do?

File: MANIFEST.in [hunk 3]
```diff
@@ -4,19 +4,32 @@ include SYMLINK_CACHE.json
 include requirements.txt
 include shippable.yml
 include bin/ansible-test
```
Comment (mattclay) [inline MANIFEST.in]: This should no longer be needed, as we should be including all files in `bin` already, correct?

File: MANIFEST.in [hunk 4]
```diff
@@ -4,19 +4,32 @@ include SYMLINK_CACHE.json
...
+recursive-include lib/ansible/module_utils/powershell *.psm1
+recursive-include lib/ansible/modules/windows *.ps1
+recursive-include lib/ansible/galaxy/data *.yml *.j2 README.md ansible.cfg inventory .git_keep
+recursive-include lib/ansible/config/ *.yml
 recursive-include licenses *
 recursive-include packaging *
-recursive-include test *
+recursive-include test/cache .keep
+recursive-include test/integration *
+recursive-include test/legacy *
```
Comment (mattclay) [inline MANIFEST.in]: We should probably omit these. They're not run in CI and are generally in an unknown state.

File: MANIFEST.in [hunk 5]
```diff
@@ -4,19 +4,32 @@ include SYMLINK_CACHE.json
...
+recursive-exclude test/lib/ansible_test *.pyc *.pyo *.bak *.orig *~ *.swp
```
Comment (mattclay) [inline MANIFEST.in]: Why do we need excludes here but not for `test/sanity` or `test/units`?

Comment (abadger) [reply MANIFEST.in]: If it's safe, I can target it at a different level but I wasn't sure of that.  sanity is probably okay.  units, there's a small chance of someone wanting to drop a data file in that uses those extensions.  integration, I think I would definitely avoid.

Things like PluginLoader tests might want to use a .pyo or .pyc file as part of a test case, for instance.

File: MANIFEST.in (hunk 3 above)
Comment (abadger) [reply MANIFEST.in]: Yep, reoved bin/ansible-test i nthe latest version

File: MANIFEST.in (hunk 1 above)
Comment (abadger) [reply MANIFEST.in]: Fixed

File: MANIFEST.in (hunk 2 above)
Comment (abadger) [reply MANIFEST.in]: <nod> Fixed.

File: MANIFEST.in (hunk 4 above)
Comment (abadger) [reply MANIFEST.in]: Legacy removed from MANIFEST.in and ignored in the package-data test now.

File: MANIFEST.in (hunk 5 above)
Comment (abadger) [reply MANIFEST.in]: mattclay and I discussed this and slimmed the MANIFEST file down to 6 directories with ```*``` wildcards.  Decided an eventual goal of removing all the wildcards would be better than blacklisting extensions.  Removed the extension blacklist in the latest version.

**Review by mattclay — APPROVED** [review]

**Comment by mattclay** [comment]
This works well enough for now. However, we should work towards eliminating excludes, since we can't predict all the possible file patterns that may pollute the source tree. It will be safer to limit ourselves to only includes. However, at least for some of the test directories this will require us to enumerate all of the special cases.

**Pull request merged**